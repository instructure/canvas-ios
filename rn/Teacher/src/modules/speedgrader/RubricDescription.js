//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

// @flow

import React, { Component } from 'react'
import { View, StyleSheet, Text } from 'react-native'
import { connect } from 'react-redux'
import i18n from 'format-message'
import Navigator from '../../routing/Navigator'
import Screen from '../../routing/Screen'
import CanvasWebView from '../../common/components/CanvasWebView'

export class RubricDescription extends Component<*> {
  renderLongDescription () {
    if (!this.props.description || this.props.description.length === 0) {
      return (
        <View style={styles.emptyState}>
          <Text style={styles.emptyStateText}>{i18n('There currently is no long description for this item.')}</Text>
        </View>
      )
    }
    return (
      <View style={styles.container}>
        <CanvasWebView html={this.props.description} automaticallySetHeight navigator={this.props.navigator}/>
      </View>
    )
  }

  render () {
    const description = this.renderLongDescription()
    return (
      <Screen
        title={i18n('Long Description')}
      >
        { description }
      </Screen>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    paddingTop: 24,
    paddingHorizontal: 16,
  },
  emptyState: {
    flex: 1,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
    paddingHorizontal: 32,
  },
  emptyStateText: {
    color: '#2D3B45',
    fontSize: 16,
    lineHeight: 19,
    textAlign: 'center',
  },
})

export function mapStateToProps (state: AppState, ownProps: RubricDescriptionOwnProps): RubricDescriptionDataProps {
  let rubric = state.entities.assignments[ownProps.assignmentID].data.rubric
  if (!rubric) {
    return { description: '' }
  }

  return {
    description: rubric.find(r => r.id === ownProps.rubricID).long_description,
  }
}

const Connected = connect(mapStateToProps)(RubricDescription)
export default (Connected: any)

type RubricDescriptionOwnProps = {
  assignmentID: string,
  rubricID: string,
  navigator: Navigator,
}

type RubricDescriptionDataProps = {
  description: string,
}
