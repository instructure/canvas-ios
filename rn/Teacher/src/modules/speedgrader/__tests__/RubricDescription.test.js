//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

// @flow

import React from 'react'
import { RubricDescription, mapStateToProps } from '../RubricDescription'
import renderer from 'react-test-renderer'

jest.unmock('react-native/Libraries/Components/ScrollView/ScrollView')

const templates = {
  ...require('../../../__templates__/helm'),
  ...require('../../../redux/__templates__/app-state'),
  ...require('../../../__templates__/rubric'),
  ...require('../../../__templates__/assignments'),
  ...require('../../../__templates__/session'),
}

let ownProps = {
  assignmentID: '1',
  rubricID: '1',
  navigator: templates.navigator(),
}

let defaultProps = {
  ...ownProps,
  description: 'A satisfactory description',
}

describe('RubricDescription', () => {
  beforeEach(() => jest.clearAllMocks())

  it('renders', () => {
    let tree = renderer.create(
      <RubricDescription {...defaultProps} />
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders without description', () => {
    let props = {
      ...defaultProps,
      description: null,
    }
    let tree = renderer.create(
      <RubricDescription {...props} />
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })
})

describe('mapStateToProps', () => {
  it('returns an empty description when there is no rubric', () => {
    let state = templates.appState({
      entities: {
        assignments: {
          '1': {
            data: templates.assignment({
              id: '1',
            }),
          },
        },
      },
    })

    let props = mapStateToProps(state, ownProps)
    expect(props.description).toEqual('')
  })

  it('returns the rubric description whent here is a rubric', () => {
    let rubric = templates.rubric({ id: '1' })
    let state = templates.appState({
      entities: {
        assignments: {
          '1': {
            data: templates.assignment({
              id: '1',
              rubric: [rubric],
            }),
          },
        },
      },
    })

    let props = mapStateToProps(state, ownProps)
    expect(props.description).toEqual(rubric.long_description)
  })
})
