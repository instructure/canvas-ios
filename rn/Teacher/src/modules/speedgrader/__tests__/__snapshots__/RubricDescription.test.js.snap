// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RubricDescription renders 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "paddingHorizontal": 16,
        "paddingTop": 24,
      }
    }
  >
    <View
      testID="web-container.view"
    >
      <CanvasWebView
        automaticallySetHeight={true}
        bounces={false}
        html="A satisfactory description"
        navigator={
          Object {
            "dismiss": [MockFunction],
            "dismissAllModals": [MockFunction],
            "isModal": false,
            "pop": [MockFunction],
            "replace": [MockFunction],
            "show": [MockFunction],
            "traitCollection": [MockFunction],
          }
        }
        onError={[Function]}
        onFinishedLoading={[Function]}
        onHeightChange={[Function]}
        onMessage={[Function]}
        onNavigation={[Function]}
        scrollEnabled={true}
        source={
          Object {
            "baseURL": "http://mobiledev.instructure.com/",
            "html": "A satisfactory description",
          }
        }
        style={
          Object {
            "flex": 1,
          }
        }
      />
    </View>
  </View>
</RCTSafeAreaView>
`;

exports[`RubricDescription renders without description 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "center",
        "paddingHorizontal": 32,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#2D3B45",
          "fontSize": 16,
          "lineHeight": 19,
          "textAlign": "center",
        }
      }
    >
      There currently is no long description for this item.
    </Text>
  </View>
</RCTSafeAreaView>
`;
